Kinvey/Kinvey/HttpRequest.swift

```diff
@@ -405,30 +410,49 @@ internal class HttpRequest<Result>: TaskProgressRequest, Request {
                     }
                 }
                 if response.statusCode == 401,
+                    retries,
                     let user = self.credential as? User
                 {
-                    guard let refreshToken = user.refreshToken else {
-                        user.logout()
-                        completionHandler?(data, HttpResponse(response: response), error)
-                        return
-                    }
-                    let options = try! Options(self.options, authServiceId: self.client.clientId)
-                    MIC.login(refreshToken: refreshToken, options: options) {
-                        switch $0 {
-                        case .success(let user):
-                            self.credential = user
-                            self.execute(urlSession: urlSession, completionHandler)
-                        case .failure(let error):
-                            if let error = error as? Kinvey.Error {
-                                switch error {
-                                case .invalidCredentials:
-                                    if let user = self.credential as? User {
-                                        user.logout()
+                    DispatchQueue.global(qos: .default).async {
+                        guard !client.refreshingToken else {
```

Comment:
Is the `client.refreshingToken` made thread-safe because it is being accessed and set from within the `DispatchQueue`?

Reply:
right, so this is a serial queue and I'm calling `wait()` on the `DispatchGroup` so the combination of both makes the entire refresh token process thread-safe